### PR TITLE
[Sitemaps] Google Sitemap PageMap extensions, implements #388

### PR DESCRIPTION
**Pull request by sebastian-nagel**
Implementation of Google's PageMap extension, see #388.

### CODE REVIEW

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import crawlercommons.sitemaps.extension.PageMapDataObject;
+
+/**
+ * Handle SAX events in the Google's Programmable Search Engine <a href=
+ * "https://support.google.com/programmable-search/answer/1628213">PageMaps</a>
+ * extension namespace.
+ */
+public class PageMapsHandler extends ExtensionHandler {
+
+    PageMap currPageMap;
```
Comment (rzo1) [inline src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:31]: Is there a specific reasons for that fields being package-private?

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    public void reset() {
+        super.reset();
+        resetCurrent();
+    }
+
+    private void resetCurrent() {
+        currAttrName = null;
+        currDataObj = null;
+        currAttrName = null;
+        currAttrVal = new StringBuilder();
```
Comment (rzo1) [inline src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:51]: Don't know if we are in a critical code path here, but if we don't want to re-create the object, a `setLength(0)` would be sufficient.

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        resetCurrent();
+    }
+
+    private void resetCurrent() {
+        currAttrName = null;
```
Comment (rzo1) [inline src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:48]: This is a duplicate. `currAttrValFromAttr`is missing here.

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java (hunk 2 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:51]: Ok.

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java (hunk 3 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:48]: Good catch! `currAttrValFromAttr` is set to null a few lines below.

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java (hunk 1 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:31]: No. Only my laziness. Will make it class-private.

**Comment by sebastian-nagel** [comment]
Thanks, @rzo1! I've addressed all your comments and also ensured that no NPEs happen with not properly nested PageMap data, eg., DataObject without surrounding PageMap.

File: src/main/java/crawlercommons/sitemaps/extension/PageMapDataObject.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        sb.append("{type = ").append(type);
+        sb.append(", id = ").append(id);
+        sb.append(", attributes = ").append(attributes);
+        sb.append("}");
+        return sb.toString();
+    }
+
+    @Override
+    public Map<String, String[]> asMap() {
+        // TODO Auto-generated method stub
```
Comment (rzo1) [inline src/main/java/crawlercommons/sitemaps/extension/PageMapDataObject.java:82]: Is this something we should fix?

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+ * Handle SAX events in the Google's Programmable Search Engine <a href=
+ * "https://support.google.com/programmable-search/answer/1628213">PageMaps</a>
+ * extension namespace.
+ */
+public class PageMapsHandler extends ExtensionHandler {
+
+    private PageMap currPageMap;
+    private PageMapDataObject currDataObj;
+    private String currAttrName;
+    private StringBuilder currAttrVal = new StringBuilder();;
```
Comment (rzo1) [inline src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java:34]: Double ";"

File: src/test/java/crawlercommons/sitemaps/extension/PageMapsTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package crawlercommons.sitemaps.extension;
```
Comment (rzo1) [inline src/test/java/crawlercommons/sitemaps/extension/PageMapsTest.java:1]: missing license header

**Review by rzo1 — APPROVED** [review]

File: src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java (hunk 5 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/sitemaps/sax/extension/PageMapsHandler.java]: Done.

File: src/test/java/crawlercommons/sitemaps/extension/PageMapsTest.java (hunk 6 above)
Comment (sebastian-nagel) [reply src/test/java/crawlercommons/sitemaps/extension/PageMapsTest.java:16]: Added.

File: src/test/resources/sitemaps/extension/pagemaps-sitemap.xml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (sebastian-nagel) [reply src/test/resources/sitemaps/extension/pagemaps-sitemap.xml:1]: Don't know, the other sample files in this directory also do not have license headers.

File: src/main/java/crawlercommons/sitemaps/extension/PageMapDataObject.java (hunk 4 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/sitemaps/extension/PageMapDataObject.java]: Yes. I've moved parts of the asMap() method of the PageMap class to here.

**Comment by rzo1** [comment]
lgtm!

**Comment by sebastian-nagel** [comment]
Squashed commits and rebased to master to include the fix for #441 affecting the automatic builds.

**Review by jnioche — APPROVED** [review]

**Comment by jnioche** [comment]
thanks @sebastian-nagel and @rzo1

**Pull request merged**